source/common/upstream/cds_api_impl.cc

```diff
@@ -51,12 +47,44 @@ void CdsApiImpl::onConfigUpdate(const ResourceVector& resources, const std::stri
   for (const auto& cluster : resources) {
     MessageUtil::validate(cluster);
   }
-  // We need to keep track of which clusters we might need to remove.
+
   ClusterManager::ClusterInfoMap clusters_to_remove = cm_.clusters();
   for (auto& cluster : resources) {
-    const std::string cluster_name = cluster.name();
+    clusters_to_remove.erase(cluster.name());
+  }
+  Protobuf::RepeatedPtrField<std::string> to_remove_repeated;
+  for (auto cluster : clusters_to_remove) {
+    *to_remove_repeated.Add() = cluster.first;
+  }
+
+  Protobuf::RepeatedPtrField<envoy::api::v2::Resource> to_add_repeated;
+  for (const auto& cluster : resources) {
+    envoy::api::v2::Resource to_add;
+    to_add.set_name(cluster.name());
+    to_add.set_version(version_info);
+    ProtobufWkt::Any any;
+    any.PackFrom(cluster);
+    *to_add.mutable_resource() = any;
+    *to_add_repeated.Add() = to_add;
+  }
+
+  onConfigUpdate(to_add_repeated, to_remove_repeated, version_info);
```

Comment:
Great, this is basically what I had in mind :) We should be careful in various stats to not break any existing behaviors, I think we're all good there, but please verify.

Reply:
Awesome. The existing CdsApiImplTest does a bit of examination of the xDS-related stats; do you think those existing tests are good enough, or should something else be added?

Reply:
Yeah, LG, I think the main thing is to consider going forward how these state-of-the-world legacy updates look like in terms of the new stats you are adding for incremental (or whether they are hidden from them).

Reply:
I don't think I've added any new incremental-specific stats. The incremental code uses the same stats as vanilla xDS.